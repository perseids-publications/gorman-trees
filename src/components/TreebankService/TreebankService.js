import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import {
  MessagingService,
  ResponseMessage,
  WindowIframeDestination as Destination,
} from 'alpheios-messaging';
import { buildQueryString } from '../../lib/params';

import ArethusaWrapper from '../ArethusaWrapper';

const config = {
  name: 'treebank',
  targetIframeID: 'string-not-used',
  targetURL: 'string-not-used',
  commModes: [Destination.commModes.RECEIVE],
};

const error = (request, message, code) => ResponseMessage.Error(request, new Error(message), code);

const redirectLink = ({ wordIds, config: c, sentenceId }) => {
  const query = buildQueryString({ w: wordIds, config: c });

  if (query === '') {
    return sentenceId;
  }

  return `${sentenceId}?${query}`;
};

class TreebankService extends Component {
  constructor(props) {
    super(props);

    this.state = {
      arethusaLoaded: false,
      redirectTo: null,
    };
    this.messageHandler = this.messageHandler.bind(this);
    this.setArethusaLoaded = this.setArethusaLoaded.bind(this);
  }

  componentDidMount() {
    this.destination = new Destination({ ...config, receiverCB: this.messageHandler });
    this.service = new MessagingService('treebank-service', this.destination);

    // eslint-disable-next-line no-undef
    window.document.body.addEventListener('ArethusaLoaded', this.setArethusaLoaded);
  }

  componentWillUnmount() {
    this.destination.deregister();

    // eslint-disable-next-line no-undef
    window.document.body.removeEventListener('ArethusaLoaded', this.setArethusaLoaded);
  }

  setArethusaLoaded() {
    this.setState({ arethusaLoaded: true });
  }

  messageHandler(request, responseFn) {
    const { arethusa } = this.props;
    const { arethusaLoaded } = this.state;
    const { body } = request;
    const [name] = Object.keys(body);

    if (!arethusaLoaded) {
      responseFn(error(request, 'Arethusa is Not Loaded', ResponseMessage.errorCodes.SERVICE_UNINITIALIZED));
      return;
    }

    try {
      switch (name) {
        case 'gotoSentence':
          this.setState({
            redirectTo: redirectLink(body.gotoSentence),
          });

          responseFn(ResponseMessage.Success(request, { status: 'success' }));
          break;
        case 'getMorph':
          responseFn(ResponseMessage.Success(
            request,
            arethusa.getMorph(body.getMorph.sentenceId, body.getMorph.wordId),
          ));
          break;
        case 'refreshView':
          responseFn(ResponseMessage.Success(request, arethusa.refreshView()));
          break;
        case 'findWord':
          responseFn(
            ResponseMessage.Success(
              request,
              arethusa.findWord(
                body.findWord.sentenceId,
                body.findWord.word,
                body.findWord.prefix,
                body.findWord.suffix,
              ),
            ),
          );
          break;
        default:
          responseFn(error(request, `Unsupported request: ${name}`, ResponseMessage.errorCodes.UNKNOWN_REQUEST));
      }
    } catch (err) {
      responseFn(ResponseMessage.Error(request, err, ResponseMessage.errorCodes.INTERNAL_ERROR));
    }
  }

  render() {
    const { redirectTo } = this.state;

    return (
      redirectTo && <Redirect to={redirectTo} />
    );
  }
}

TreebankService.propTypes = {
  arethusa: PropTypes.instanceOf(ArethusaWrapper).isRequired,
};

export default TreebankService;
