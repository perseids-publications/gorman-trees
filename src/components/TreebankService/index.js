import TreebankService from './TreebankService';

export default TreebankService;
