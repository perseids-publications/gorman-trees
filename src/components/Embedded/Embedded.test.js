import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import Embedded from './Embedded';
import TreebankService from '../TreebankService';

import config from './config.test.json';

it('renders an embedded publication', () => {
  const component = (
    <MemoryRouter initialEntries={['/embed/on-the-murder-of-eratosthenes-1-50/1']}>
      <Embedded config={config} />
    </MemoryRouter>
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});

it('provides an API to communicate with an embedded publication', () => {
  const component = (
    <MemoryRouter initialEntries={['/embed/on-the-murder-of-eratosthenes-1-50/1']}>
      <Embedded config={config} />
    </MemoryRouter>
  );
  const tree = renderer.create(component);
  const { instance: { messageHandler } } = tree.root.findByType(TreebankService);

  window.document.body.dispatchEvent(new window.Event('ArethusaLoaded'));

  messageHandler(
    { ID: 'test', body: { gotoSentence: { sentenceId: '5' } } },
    () => {},
  );

  expect(tree).toMatchSnapshot();
});

it('the API can be used to select words', () => {
  const component = (
    <MemoryRouter initialEntries={['/embed/on-the-murder-of-eratosthenes-1-50/1']}>
      <Embedded config={config} />
    </MemoryRouter>
  );
  const tree = renderer.create(component);
  const { instance: { messageHandler } } = tree.root.findByType(TreebankService);

  window.document.body.dispatchEvent(new window.Event('ArethusaLoaded'));

  messageHandler(
    { ID: 'test', body: { gotoSentence: { sentenceId: '5', wordIds: ['12'] } } },
    () => {},
  );

  expect(tree).toMatchSnapshot();
});

it('renders 404 when publication not found', () => {
  const component = (
    <MemoryRouter initialEntries={['/embed/unknown']}>
      <Embedded config={config} />
    </MemoryRouter>
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});
