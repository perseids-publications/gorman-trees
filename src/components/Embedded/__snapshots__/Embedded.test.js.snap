// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provides an API to communicate with an embedded publication 1`] = `
<div>
  <div
    className="__artsa"
  >
    <div
      className="treebankContainer"
      id="treebank_container"
    />
  </div>
  <div
    className="links"
  >
    <a
      href="https://www.example.com/on-the-murder-of-eratosthenes-1-50/5"
      rel="noopener noreferrer"
      target="_blank"
    >
      Credits and more information
    </a>
  </div>
  <link
    href="https://www.example.com/arethusa/css/arethusa.min.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/colorpicker.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/font-awesome.min.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/foundation-icons.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/widget.css"
    rel="stylesheet"
    type="text/css"
  />
</div>
`;

exports[`renders 404 when publication not found 1`] = `
<div
  className="container pt-5"
>
  <div
    className="row col-12 pt-5 pb-3"
  >
    <div
      className="col-12 text-center"
    >
      <h1>
        Error 404
      </h1>
    </div>
  </div>
  <div
    className="row col-12 pb-3"
  >
    <div
      className="col-12 text-center"
    >
      <h2>
        Publication not found
      </h2>
    </div>
  </div>
</div>
`;

exports[`renders an embedded publication 1`] = `
<div>
  <div
    className="__artsa"
  >
    <div
      className="treebankContainer"
      id="treebank_container"
    />
  </div>
  <div
    className="links"
  >
    <a
      href="https://www.example.com/on-the-murder-of-eratosthenes-1-50/1"
      rel="noopener noreferrer"
      target="_blank"
    >
      Credits and more information
    </a>
  </div>
  <link
    href="https://www.example.com/arethusa/css/arethusa.min.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/colorpicker.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/font-awesome.min.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/foundation-icons.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/widget.css"
    rel="stylesheet"
    type="text/css"
  />
</div>
`;

exports[`the API can be used to select words 1`] = `
<div>
  <div
    className="__artsa"
  >
    <div
      className="treebankContainer"
      id="treebank_container"
    />
  </div>
  <div
    className="links"
  >
    <a
      href="https://www.example.com/on-the-murder-of-eratosthenes-1-50/5"
      rel="noopener noreferrer"
      target="_blank"
    >
      Credits and more information
    </a>
  </div>
  <link
    href="https://www.example.com/arethusa/css/arethusa.min.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/colorpicker.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/font-awesome.min.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/foundation-icons.css"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="https://www.example.com/arethusa/css/widget.css"
    rel="stylesheet"
    type="text/css"
  />
</div>
`;
